Add Debug mode documentation (#745)
gravitee-io/gravitee-docs: pages/apim/3.x/user-guide/publisher/design-studio/design-studio-debug-mode.adoc

* Understand which policies are triggered by a specific request (including Platform-level policies).
* Visualise which policies are executed or skipped (conditional policy).
* Understand the order of execution of your policies.
* Trace transformations and changes made to headers, body and execution context.
* Easily troubleshoot and find root cause of errors, misbehaviors or performance issues.

== How to Get Started With Debug Mode

. link:{{ '/apim/3.x/apim_publisherguide_design_studio_create.html' | relative_url }}[Design your API^].
. Select *Debug* tab.
. Configure your request.
. Select *Send*.

== How To Test Your API Concept with Debug Mode

You can send a request to your *API* endpoint through the embedded HTTP client available under the *Debug* tab.

NOTE: You don't need to save or redeploy the API to test your changes when using the Debug Mode.

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-example-empty.png %}[Debug Mode empty example]

> The embedded HTTP client let's you configure different attributes of your request:

Reviewer: ```suggestion
The embedded HTTP client let you configure different attributes of your request:
```

* *HTTP method* to select the HTTP verb (e.g., `GET`, `POST`, `PUT`, etc.).
* *Path* to test a specific subpath and provide query parameters. 

NOTE: The context-path of the API your are debugging is already taken into account. Path refers to the subpath: myapi-context-path/*path*?queryparam=value

* *Headers* to provide a custom list of HTTP request headers.
* *Request body* to configure a custom request body. 

> Since debug mode emulates an external call to your API, it will resolve the flow as the gateway would do, so if you want to debug a specific flow you must configure your request (path, headers) to replicate the conditions at which the flow is triggered.

Reviewer: ```suggestion
Since debug mode emulates an external call to your API, it will resolve the flow as the gateway would do, so if you want to debug a specific flow you must configure your request (path, headers) to replicate the conditions on which the flow is triggered.
```

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-example.png %}[Debug Mode example]

The debug mode displays all policies in a vertical timeline so you can easily visualize the exact order of execution.

=== Different state for policies :

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-legend.png %}[Debug Mode legend]

* *executed* - the policy has been executed properly

> * *skipped* - the policy contains a condition that has been fulfilled. Refer to the input/output inspector for more details on the evaluation of the condition.

Reviewer: I'm assuming the condition should **NOT** be fulfilled if we skip the execution. am I right ?

* *error* - an error occurred during policy execution. Refer to the input/output inspector for more details on the error.

By selecting a specific policy in the timeline, you have access to additional information regarding the input/output of the policy:

* header
* context attributes
* body

The inspector relies on 3 colors to indicate the nature of the change:

* *green* color indicates a addition
* *orange* color indicates a edition
* *red* color indicates a deletion

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-inspector.png %}[Debug Mode inspector]

=== Timeline

==== How to interpret ?

> The order in which the policies appear in the timeline reflects the exact order in which they have been executed by the gatewau at runtime. 

Reviewer: ```suggestion
The order in which the policies appear in the timeline reflects the exact order in which they have been executed by the gateway at runtime. 
```

> Note that *this order may differ* from the order in which policies where placed in the Policy Studio during the design phase. 

Reviewer: ```suggestion
Note that this order *may* differ from the order in which policies where placed in the Policy Studio during the design phase. 
```

This is due to a performance optimization applied at runtime on the the policy chain. The gateway always executes policies interacting with the HTTP Header part of the request (onRequest, onResponse) before policies interacting with the body part of the request (onRequestContent, onResponseContent).

Also, a policy may appear twice in the timeline if it interacts with both the head and the body part of the request.

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-policy-chain.png %}[Policies execution chain order]

==== How to navigate ?

> You can scroll through the list of policies via the timeline. You can also quickly access a specific policy by clicking on a specific policy in the quick access timeline.

Reviewer: ```suggestion
You can scroll through the list of policies via the timeline. You can also quickly access a specific policy by clicking on a it in the quick access timeline.
```

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-timeline.png %}[Debug Mode timeline]

> By clicking *Request Input* or *Request Output*, you can view the global transformation on your request and difference between what's have been received by the gateway and what's sent to your backend.

Reviewer: ```suggestion
By clicking *Request Input* or *Request Output*, you can view the global transformation on your request and difference between what has been received by the gateway and what has been sent to your backend.
```

> By clicking *Response Input* or *Response Output*, you can view the global transformation on your response and difference between what's been received from the backend and what's sent back to your client app.

Reviewer: ```suggestion
By clicking *Response Input* or *Response Output*, you can view the global transformation on your response and difference between what has been received from the backend and what has been sent back to your client app.
```

* Understand which policies are triggered by a specific request (including Platform-level policies).
* Visualise which policies are executed or skipped (conditional policy).
* Understand the order of execution of your policies.
* Trace transformations and changes made to headers, body and execution context.
* Easily troubleshoot and find root cause of errors, misbehaviors or performance issues.

== How to Get Started With Debug Mode

. link:{{ '/apim/3.x/apim_publisherguide_design_studio_create.html' | relative_url }}[Design your API^].
. Select *Debug* tab.
. Configure your request.
. Select *Send*.

== How To Test Your API Concept with Debug Mode

You can send a request to your *API* endpoint through the embedded HTTP client available under the *Debug* tab.

NOTE: You don't need to save or redeploy the API to test your changes when using the Debug Mode.

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-example-empty.png %}[Debug Mode empty example]

> The embedded HTTP client, let's you can configure different attributes of your request:

Author: @allandenis38 not sure about the `let's` ?

* *HTTP method* to select the HTTP verb (e.g., `GET`, `POST`, `PUT`, etc.).
* *Path* to test a specific subpath and provide query parameters. 

NOTE: The context-path of the API your are debugging is already taken into account. Path refers to the subpath: myapi-context-path/*path*?queryparam=value

* *Headers* to provide a custom list of HTTP request headers.
* *Request body* to configure a custom request body. 

Since debug mode emulates an external call to your API, it will resolve the flow as the gateway would do, so if you want to debug a specific flow you must configure your request (path, headers) to replicate the conditions at which the flow is triggered.

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-example.png %}[Debug Mode example]

The debug mode displays all policies in a vertical timeline so you can easily visualize the exact order of execution.

=== Different state for policies :

image:{% link images/apim/3.x/api-publisher-guide/design-studio/debug-mode/debug-mode-legend.png %}[Debug Mode legend]

* *executed* - the policy has been executed properly

> * *skipped* - the policy is contains a condition that has been fulfilled. Refer to the information panel for more details on the evaluation of the condition.

Author: the policy contains ?
